File: fapla-js/Scope.js
/**
 * Created by amin on 12/30/16.
 */
"use strict";
var Scope = (function () {
    function Scope(name, parent) {

        this.name = name;
        this.table = {};
        this.parent = parent;

        this.enterScope = function () {
            return new Scope(null, this);
        }

        this.exitScope = function () {
            return this.parent;
        }

        this.addSymbol = function (symbol) {
            if(!this.findSymbol(symbol.name)) this.table[symbol.name] = symbol;
            else return false;
            return true;
        }

        this.findSymbol = function (symbolName) {
            var s = this.table[symbolName];
            if (s) return s;
            else if (this.parent) return this.parent.findSymbol(symbolName);
            else return null;
        }
    }

    return Scope;
}());
exports.Scope = Scope;

File: fapla-java/src/Scope.java
import java.util.Hashtable;


public class Scope {

    private Hashtable<String, Symbol> symboleHashtable;
    private Scope parent;

    public Scope(Scope parent) {
        this.parent = parent;
        this.symboleHashtable = new Hashtable<>();
    }

    public Scope enterScope() {
        return new Scope(this);
    }

    public Scope exitScope() {
        return this.parent;
    }

    public void addSymbol(Symbol symbol) {
        symboleHashtable.put(symbol.name, symbol);
    }

    public Symbol findSymbol(String symbolName) {
        Symbol b = symboleHashtable.get(symbolName);
        if (b != null) return b;
        else if (parent != null) return parent.findSymbol(symbolName);
        return null;
    }

    public Hashtable<String, Symbol> getSymboleHashtable() {
        return symboleHashtable;
    }
}
